### PR TITLE
Refactor DesignerHost to use Dictionary<IComponent, IDesigner> instead of Hashtable

### PR DESCRIPTION
**Pull request by Jericho**
Related: #8143 

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/8272)

### CODE REVIEW

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.cs [hunk 1]
```diff
@@ -688,9 +688,12 @@ private void Unload()
...
             _surface?.OnUnloading();
 
-            if (GetService(typeof(IHelpService)) is IHelpService helpService && _rootComponent is not null && _designers[_rootComponent] is not null)
+            if (GetService(typeof(IHelpService)) is IHelpService helpService
+                && _rootComponent is not null
+                && _designers.TryGetValue(_rootComponent, out IDesigner designer)
+                && designer is not null)
             {
-                helpService.RemoveContextAttribute("Keyword", "Designer_" + _designers[_rootComponent].GetType().FullName);
+                helpService.RemoveContextAttribute("Keyword", "Designer_" + designer.GetType().FullName);
```
Comment (dreddy-work) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.cs:696]: nit: please use string interpolation wherever possible.

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.cs [hunk 2]
```diff
@@ -1138,7 +1139,11 @@ IDesigner IDesignerHost.GetDesigner(IComponent component)
         {
             ArgumentNullException.ThrowIfNull(component);
 
-            return _designers[component] as IDesigner;
+            // It's important NOT to use "_designers[component]" here because we want to
```
Comment (dreddy-work) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.cs:1142]: I do not see a value add of this comment after we merge it.  I may just add "Return null if designer isn't available".

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.cs (hunk 1 above)
Comment (Jericho) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.cs]: Done

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.cs (hunk 2 above)
Comment (Jericho) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerHost.cs]: Comment removed

**Review by JeremyKuhne — APPROVED** [review]

**Review by lonitra — APPROVED** [review]
Thank you!

**Pull request merged**